### PR TITLE
Migrate Gemma3VisionEncoderLayer and VisionEmbedder to NNX

### PR DESCRIPTION
**Pull request by hengtaoguo**
# Description

This PR migrates the `Gemma3VisionEncoderLayer` and `VisionEmbedder` class to use NNX.
- All submodules migrated to NNX. Of note, we also migrated from `nnx.Dense/nnx.MultiHeadAttention` to MaxText `DenseGeneral/Attention`: MaxText format migration complete. 
- Preserve the checkpoint's current parameter names and hierarchy.
- The `nnx_wrappers.to_linen` bridge is used within `encoders.py`.
- Refactored hard-coded Gemma3 multimodal configuration to `base.yml` for clarity.
- Including @aireenmei's fix on the `precision=self.config.matmul_precision` across nnx.Conv/DenseGeneral layers. Adding precision flags to `attention_op.py` for einsum, and this change drastically reduces KL divergence from ~8.7 to ~0.03.

# Tests

Tested with multimodal decode command:
```
python -m MaxText.decode MaxText/configs/base.yml model_name=gemma3-4b tokenizer_path=assets/tokenizer.gemma3 load_parameters_path=gs://maxtext-gemma/unified/gemma3/4b/unscanned/2025-08-09-01-17/0/items per_device_batch_size=1 run_name=ht_test max_prefill_predict_length=272 max_target_length=300 steps=1 async_checkpointing=false scan_layers=false use_multimodal=true prompt=\'Describe\ image\ \<start_of_image\>\' image_path=\'MaxText/test_assets/test_image.jpg\' attention=\'dot_product\'
```
Results:
```
Input `<start_of_turn>user
Describe image <start_of_image><end_of_turn>
<start_of_turn>model
` -> `Here's a description of the image:

**Overall Impression:** The image is a bright, expansive cityscape view of Seattle, Washington, with`
```

Tested with sft_trainer with [command](https://paste.googleplex.com/5858533295521792) and [workload](https://pantheon.corp.google.com/logs/query;query=resource.type%3D%22k8s_container%22%0Aresource.labels.project_id%3D%22cloud-tpu-multipod-dev%22%0Aresource.labels.location%3D%22europe-west4%22%0Aresource.labels.cluster_name%3D%22v5p-8-bodaborg-europe-west4-b%22%0Aresource.labels.namespace_name%3D%22default%22%0Aresource.labels.pod_name:%22hengtaoguo-2025-09-04-17-14-36-slice-job-0-0-%22%20severity%3E%3DDEFAULT%0Atimestamp%20%3E%3D%20%222025-09-04T17:16:17.719938761Z%22%20AND%20timestamp%20%3C%3D%20%222025-09-04T17:24:42.873411327Z%22;storageScope=project;cursorTimestamp=2025-09-04T17:16:40.092655322Z;duration=P1D?e=13802955&mods=allow_workbench_image_override&project=cloud-tpu-multipod-dev)

KL divergence check [command](https://paste.googleplex.com/6201259405213696):
```
max KL divergence = 0.030590716749429703
```

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

**Comment by hengtaoguo** [comment]
> Generally LGTM, thank you @hengtaoguo. I see the decode command in the PR description. Did you also test that train works properly?

Thanks for the review! Just updated the sft_train test run in the description, with [command](https://paste.googleplex.com/5858533295521792) and [workload](https://pantheon.corp.google.com/logs/query;query=resource.type%3D%22k8s_container%22%0Aresource.labels.project_id%3D%22cloud-tpu-multipod-dev%22%0Aresource.labels.location%3D%22europe-west4%22%0Aresource.labels.cluster_name%3D%22v5p-8-bodaborg-europe-west4-b%22%0Aresource.labels.namespace_name%3D%22default%22%0Aresource.labels.pod_name:%22hengtaoguo-2025-09-04-17-14-36-slice-job-0-0-%22%20severity%3E%3DDEFAULT%0Atimestamp%20%3E%3D%20%222025-09-04T17:16:17.719938761Z%22%20AND%20timestamp%20%3C%3D%20%222025-09-04T17:24:42.873411327Z%22;storageScope=project;cursorTimestamp=2025-09-04T17:16:40.092655322Z;duration=P1D?e=13802955&mods=allow_workbench_image_override&project=cloud-tpu-multipod-dev).

**Review by bvandermoon — APPROVED** [review]
Thanks @hengtaoguo. LGTM. One other thing that would be good is to check that the memory usage is the same before/after this change

**Comment by hengtaoguo** [comment]
> Thanks @hengtaoguo. LGTM. One other thing that would be good is to check that the memory usage is the same before/after this change

Thanks for the reminder! After [checking](https://screenshot.googleplex.com/Ajx8wXxUkBoFizc), the old linen version consumes ~5.18G and new NNX version consumes ~4.82G, decreased a little after the migration.

**Comment by bvandermoon** [comment]
> > Thanks @hengtaoguo. LGTM. One other thing that would be good is to check that the memory usage is the same before/after this change
> 
> Thanks for the reminder! After [checking](https://screenshot.googleplex.com/Ajx8wXxUkBoFizc), the old linen version consumes ~5.18G and new NNX version consumes ~4.82G, decreased a little after the migration.

This is surprising. Could you please collect a profile so we can check the memory?

File: src/MaxText/layers/gemma3.py [hunk 1]
```diff
@@ -290,159 +296,196 @@ def _posemb_sincos_2d(
...
 
-class Einsum(nn.Module):
+class Einsum(nnx.Module):
   """Einsum is a convenience module for parameterized tensor multiplication."""
 
-  shape: tuple[int, ...]
-  weight_name: str = "w"
-  initializer: nn.initializers.Initializer = nn.initializers.normal()
-  dtype: jnp.dtype | None = None
+  def __init__(
```
Comment (aireenmei) [inline src/MaxText/layers/gemma3.py:434]: Could you add a precision param to this module and pass config.matmul_precision when calling it? Similar to my previous change: https://github.com/AI-Hypercomputer/maxtext/compare/main...aireen/gemma3-precision

File: src/MaxText/layers/gemma3.py [hunk 2]
```diff
@@ -467,92 +512,115 @@ def __call__(self, x):
...
-          name,
-          nn.initializers.normal(stddev=1 / (width**0.5)),
-          (1, shape_product, width),
-          dtype,
-      )
+      shape = (1, seqshape[0] * seqshape[1], width)
+      initializer = nnx.initializers.normal(stddev=1 / (width**0.5))
+      return nnx.Param(initializer(self.rngs.params(), shape, dtype))
     elif typ == "sincos2d":
       return _posemb_sincos_2d(*seqshape, width=width, dtype=dtype)
```
Comment (aireenmei) [inline src/MaxText/layers/gemma3.py]: Could you add precision to this module? Similar to my previous change in https://github.com/AI-Hypercomputer/maxtext/compare/main...aireen/gemma3-precision
